### PR TITLE
ChatX

### PR DESCRIPTION
**Pull request by guillaumearm**
## Demo
You can see a demo here : https://guillaumearm.github.io/demo-chatx

## Major features
- create a room
- remove a room (if user is the author of this room)
- select a room
- join/leave a room
- can read last 10 messages of the room when entering
- can read new messages
- can write and send new messages
- user can remove his messages
- author of a room can remove any messages
- admin can do everything

## Minor features
- select first room when entering in chat
- join and select a room at the creation
- focus the message prompt when typing
- auto scroll to the bottom of messages when receive a new message
- scrollable rooms list, messages list and users list
- author of a room has a `glyphicon-king` icon
- normal user has a `glyphicon-user` icon

#### Download
```bash
$ git clone https://github.com/guillaumearm/demo-chatx.git
$ cd demo-chatx
```

#### Installation
```bash
$ npm install
# or
$ yarn install
```

#### Use with your own firebase database (optional)
1. Please change `src/config.js` according to your database configuration
2. You should find a `firebase-rules.json` file in this project,
Please add this rules in your firebase console (https://console.firebase.google.com).

#### Tests
```bash
$ npm run lint && npm run test
```

#### Development
This will automatically launch app in development
```bash
$ npm run start
```

### Build for production
This will build assets in `dist/` folder
```bash
$ npm run build
```

### CODE REVIEW

File: src/sagas/chatx/index.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    const joinedRooms = yield call(api.GetValue, '/joinedRooms');
+    yield put(refreshJoinedRooms(joinedRooms || {}));
+
+    const tasks = yield all([
+      fork(messagesSaga),
+      fork(usersSaga),
+      fork(roomsSaga),
+    ]);
+
+    yield take(leaveChat);
```
Comment (magne4000) [inline src/sagas/chatx/index.js:22]: Proper way to ensure that you can't have multiple `enterChat` at the same time 👍

File: src/sagas/chatx/rooms/selectRoom.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      message => removedMessage({ room: roomId, message }),
+    )),
+    fork(subscribeAndDispatch(
+      `/messages/${roomId}`,
+      'child_changed',
+      message => changedMessage({ room: roomId, message }),
+      q => q.orderByChild('createdAt').limitToLast(NB_MESSAGES)
+    )),
+  ]);
+}
```
Comment (magne4000) [inline src/sagas/chatx/rooms/selectRoom.js:43]: :+1: clean way to map firebase events to redux actions with redux-saga

File: src/components/chat/RoomCreator/index.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  onSubmitForm = (e) => {
+    e.preventDefault();
+    const roomName = this.state.roomName.trim().toLowerCase();
+    if (roomName) {
+      this.setState(assoc('roomName', ''));
+      this.props.onCreate(roomName);
+    }
+  }
```
Comment (magne4000) [inline src/components/chat/RoomCreator/index.js:24]: FYI, auto-binding with this syntax is great, but has some caveats, like the disappearance of `this` context while using a library that was using this syntax + preprocessors (babel + typescript). But it's fine in most cases

File: src/reducers/chatxReducer/currentRoom.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { handleActions } from 'redux-actions';
+
+import { AUTH_LOGGED_OUT_SUCCESS } from '../../actions/actionTypes';
+import { selectRoom, leaveChat } from '../../actions/chatxActions';
+
+const initialState = null;
+
+const currentRoom = handleActions({
+  [selectRoom]: (state, action) => action.payload,
```
Comment (magne4000) [inline src/reducers/chatxReducer/currentRoom.js:9]: Without strong typings, it's hard to guess what's the payload structure.

File: src/store/commonMiddlewares.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import thunk from 'redux-thunk';
+import { routerMiddleware } from 'react-router-redux';
+import { hashHistory } from "react-router";
+
+import sagaMiddleware from '../sagas/middleware';
+
+export default [
+  thunk,
```
Comment (magne4000) [inline src/store/commonMiddlewares.js:8]: I didn't saw anywhere you using `redux-thunk`, did I miss it ?
As you probably guessed, `redux-saga` totally replaces `redux-thunk`, and is more predictable.

File: src/api/firebase.js [hunk 6]
```diff
@@ -30,18 +31,35 @@ class FirebaseApi {
...
+
+  static GetValue = async (path) => {
+    const snapshot = await firebase.database().ref(path).once('value');
+    return snapshot.val();
+  }
+
+  static Subscribe = (path, event, prepareQuery = identity) => emit => {
+    const query = prepareQuery(firebase.database().ref(path));
+    const cb = (snapshot) => {
+      emit({ ...snapshot.val(), id: snapshot.key });
```
Comment (magne4000) [inline src/api/firebase.js:55]: :+1:

**Review by magne4000 — APPROVED** [review]
- Nice semantic commit messages, we perhaps should leverage this on our side
- I hadn't seen any usage of `ramda` before, it seems pretty neat. I would like to discuss around the subject `ramda` vs `ImmutableJS` :)
- We haven't done it yet, but testing sagas seems pretty easy (and readable), so glad you did it
- Now that we are using Typescript, I feel that using action payload without strong typings is really not something I want to go back to

File: src/sagas/utils.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+import api from '../api/firebase';
+
+export const createSubscriptionChan = compose(eventChannel, api.Subscribe);
+
+export const subscribeAndDispatch = (path, event, actionCreator, prepareQuery) => (
+  function* saga() {
+    const chan = yield call(createSubscriptionChan, path, event, prepareQuery);
+    try {
+      while (true) {
```
Comment (magne4000) [inline src/sagas/utils.js:13]: You could leverage `takeEvery` here

File: src/store/commonMiddlewares.js (hunk 5 above)
Comment (guillaumearm) [reply src/store/commonMiddlewares.js:8]: `redux-thunk` is used by authentification actions, it was already there in the boilerplate.

File: src/components/chat/RoomCreator/index.js (hunk 3 above)
Comment (guillaumearm) [reply src/components/chat/RoomCreator/index.js:24]: Hum, I'm curious here, we shall talk about this

**Pull request merged**